s3-migrate/migrate.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env python
+
+import base64
+import hashlib
+import os
+import random
+import shutil
+import tempfile
+import threading
+from queue import Empty, Queue
+
+import boto3
+import dj_database_url
+import pymogilefs
+import pymysql
+import requests
+from botocore.exceptions import ClientError
+
+
+def hash_fileobj(fileobj, hasher):
+    """Efficiently hash a file object using the provided hasher."""
+    blocksize = 65536
+    fileobj.seek(0)
+    block = fileobj.read(blocksize)
+    while block:
+        hasher.update(block)
+        block = fileobj.read(blocksize)
+    fileobj.seek(0)
+    return hasher
+
+
+def md5_fileobj_hex(fileobj):
+    """Efficiently calculate the MD5 sum for a file object. Returns a hex
+string."""
+    return hash_fileobj(fileobj, hashlib.md5()).hexdigest()
+
+
+def md5_fileobj_b64(fileobj):
+    """Efficiently calculate the MD5 sum for a file object. Returns a
+Base64 encoded string."""
+    return base64.b64encode(
+        hash_fileobj(fileobj, hashlib.md5()).digest()
+    ).decode("utf-8")
+
+
+def sha1_fileobj_hex(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object. Returns a hex
+string."""
+    return hash_fileobj(fileobj, hashlib.sha1()).hexdigest()
+
+
+def sha1_fileobj_b64(fileobj):
+    """Efficiently calculate the SHA1 sum for a file object. Returns a
+Base64 encoded string."""
+    return base64.b64encode(
+        hash_fileobj(fileobj, hashlib.sha1()).digest()
+    ).decode("utf-8")
+
+
+class MyThread(threading.Thread):
+    """Thread worker with s3 and mogile clients available."""
+
+    def __init__(self, q: Queue, *args, **kwargs):
+        super(MyThread, self).__init__(*args, **kwargs)
+        self.queue = q
+        self.mogile_client = pymogilefs.client.Client(
+            trackers=os.environ['MOGILE_TRACKERS'].split(','),
+            domain='plos_repo')
+        self.s3_client = boto3.resource('s3')
+
+    def run(self):
+        while True:
+            try:
+                mogile_file = self.queue.get_nowait()
+                print(f"Migrating {mogile_file.fid} to "
+                      f"{mogile_file.make_contentrepo_key()}")
+                mogile_file.migrate(
+                    self.mogile_client,
+                    self.s3_client,
+                    os.environ['BUCKET'])
+                self.queue.task_done()
+            except Empty:
+                break
+
+    @classmethod
+    def start_pool(cls, queue):
+        """Run threads on the data provided in the queue. Return a list of
+threads to pass to `finish_pool` later."""
+        threads = []
+        for _ in range(20):
+            thread = cls(queue)
+            threads.append(thread)
+            thread.start()
+        return threads
+
+    @staticmethod
+    def finish_pool(queue, threads):
+        """Wait until queue is empty and all threads have finished
+processing."""
```

Comment:
It's not really important, but it looks like your autoformatter might have wrapped this against your intention.

You could perhaps convert a long single-line docstring into the multiline format to appease both PEP 257 and PEP 8's line limit:

```python
    def finish_pool(queue, threads):
        """
        Wait until queue is empty and all threads have finished processing.
        """
```

Reply:
No, I think I did that on purpose. Is the docstring bad?

Reply:
ehh, well, it makes `pydocstyle` mad:

![migrate py — s3-migrate 2019-11-08 15-36-03](https://user-images.githubusercontent.com/916368/68509029-fed73a00-023d-11ea-8bfe-690aecfd4d09.png)

Reply:
anyways, this is totally unimportant. Disregard if you don't care about PEP 257

Reply:
I'm using this ticket to learn all i can about python style, so I'd like to fix.

Reply:
ok, cool! Well in that case, if I were you, I'd install pydocstyle (or pylama which contains it) and get rid of those wiggles.

Reply:
Fixed.